Api pagination

will-moore opened the pull request:
# What this PR does

Updates pagination to use ```?offset=200``` to show page 2 of results instead of ```?page=2```.
We also count the total number of objects that would be returned by all "list objects" queries and return this in a new ```"meta"``` json object, along with other pagination info (current limit, offset & max limit):
E.g. ```/images/```
```
"meta": {
    "totalCount": 16739,
    "limit": 200,
    "maxLimit": 500,
    "offset": 0
},
```
This PR also adds new config settings for the default api "limit" and the "max limit" in a new ```api_settings.py```. Other api settings have also been moved to this file.

# Testing this PR

1. Go to any top-level object listing, e.g. /projects/ or /images/ and check for ```"meta"``` in json.
    - Check that 'totalCount' is correct (same as in clients)
2. Filter E.g. filter images by Dataset / orphaned and check that totalCount is correct.
3. Try using ```?offset=100``` and / or ```?limit=50```. Check that these values are shown in "meta" (and that they do pagination as expected)
4. Test the max limit by trying a limit above 500. E.g. use ```?limit=1000```. Should see that the actual limit used in the query is restricted to the max_limit (500).
5. Configure the default limit ```bin/omero config set omero.web.api.limit 300```
6. Configure the max limit ```bin/omero config set omero.web.api.max_limit 1000```
7. Check that these are used correctly in the scenarios above.

# Related reading

Discussed at web meeting [2017-01-26](https://docs.google.com/document/d/18ep65grhBURv6CcZLGVmvKQvP3RjfuyIhE288AAuFTM/edit#heading=h.9u06hgeemoyv) and [2017-02-08](https://docs.google.com/document/d/18ep65grhBURv6CcZLGVmvKQvP3RjfuyIhE288AAuFTM/edit#heading=h.3ekyh84japrw)

will-moore committed:
Use ?offset=200 parameter instead of page=2

will-moore committed:
Blitz conn.buildCountQuery() based on conn.buildQuery()

will-moore committed:
api_query.py uses conn.buildCountQuery() to add 'totalCount' to 'meta':{}

will-moore committed:
Test 'meta' dict for listing Datasets & Plates

will-moore committed:
Test 'meta' dict for listing Images, orphans

will-moore committed:
Added omero.web.max_page_size 500 to settings.py

will-moore committed:
Limit page size to MAX_PAGE

will-moore committed:
flake8 fix

snoopycrimecop commented:
Conflicting PR. Removed from build [OMERO-DEV-merge-push#584](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/584/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/584/consoleText) for more details.
Possible conflicts:
  - PR #5047 will-moore 'Api wells'
    - components/tools/OmeroWeb/test/integration/test_api_images.py
    - components/tools/OmeroWeb/test/integration/test_api_containers.py
  - PR #5058 will-moore 'Api child count'
    - components/tools/OmeroWeb/test/integration/test_api_images.py
    - components/tools/OmeroWeb/test/integration/test_api_containers.py

will-moore committed:
Moving api settings to api_settings.py

snoopycrimecop commented:
Conflicting PR. Removed from build [OMERO-DEV-merge-push#585](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/585/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/585/consoleText) for more details.
Possible conflicts:
  - PR #4991 aleksandra-tarkowska 'remove django 1.6 related workarounds'
    - components/tools/OmeroWeb/omeroweb/settings.py
  - PR #5047 will-moore 'Api wells'
    - components/tools/OmeroWeb/test/integration/test_api_images.py
    - components/tools/OmeroWeb/test/integration/test_api_containers.py
  - PR #5058 will-moore 'Api child count'
    - components/tools/OmeroWeb/test/integration/test_api_images.py
    - components/tools/OmeroWeb/test/integration/test_api_containers.py
    - components/tools/OmeroWeb/omeroweb/settings.py

will-moore committed:
Merge remote-tracking branch 'origin/develop' into api_pagination

will-moore committed:
Fix merge conflict

will-moore committed:
Don't need to remove opts['child_count'] in buildCountQuery()

will-moore committed:
Fix test_api_images.py to use api_settings and MAX_LIMIT

will-moore committed:
Fix API_VERSIONS to '0' from merge conflict

snoopycrimecop commented:
Conflicting PR. Removed from build [OMERO-DEV-merge-push#593](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/593/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/593/consoleText) for more details.
Possible conflicts:
  - Conflict detection failed
    - components/tools/OmeroWeb/test/integration/test_api_images.py

jburel commented on components/tools/OmeroPy/src/omero/gateway/__init__.py:
```diff
@@ -3181,6 +3181,36 @@ def buildQuery(self, obj_type, ids=None, params=None, attributes=None,
+        unnecessary objects to query.
+        We return just the query and omero.sys.ParametersI for the query.
+
+        :param obj_type:    Object type. E.g. "Project" see above
```
typo ``e.g.``

jburel commented on components/tools/OmeroPy/src/omero/gateway/__init__.py:
```diff
@@ -3181,6 +3181,36 @@ def buildQuery(self, obj_type, ids=None, params=None, attributes=None,
+        :param opts:        Dict of options for filtering by
+                            offset, limit and owner for all objects.
+                            Additional opts handled by _getQueryString()
+                            E.g. filter Dataset by 'project'
```
same

jburel commented on components/tools/OmeroWeb/omeroweb/api/decorators.py:
```diff
@@ -27,7 +27,8 @@
 from functools import update_wrapper
-from api_exceptions import NotFoundError, BadRequestError, CreatedObject
+from api_exceptions import NotFoundError, BadRequestError, CreatedObject, \
+    MethodNotSupportedError
```
order them

will-moore committed:
Merge remote-tracking branch 'origin/develop' into api_pagination

will-moore committed:
flake8 fixes from merge conflict

will-moore committed:
Tiny fix to avoid merge conflict

jburel commented on components/tools/OmeroPy/src/omero/gateway/__init__.py:
```diff
@@ -3181,6 +3181,36 @@ def buildQuery(self, obj_type, ids=None, params=None, attributes=None,
+                            E.g. filter Dataset by 'project'
+        :return:            (query, params)
+        """
+        # We disable pagination since we want to count ALL results
```
something to test for large number of results returned

jburel commented on components/tools/OmeroWeb/omeroweb/api/api_settings.py:
```diff
@@ -0,0 +1,57 @@
+
+# For any given release of api, we may support
+# one or more versions of the api.
+# E.g. /api/v1.0/
```
typo

jburel commented on components/tools/OmeroWeb/test/integration/test_api_projects.py:
```diff
@@ -395,17 +395,16 @@ def test_marshal_projects_pagination(self, user1, user2,
         # Check that page 2 gives next 2 projects
-        page = 2
-        payload = {'limit': limit, 'page': page}
+        payload = {'limit': limit, 'offset': limit}
```
just to avoid possible mistake or miss some errors, I will use different values e.g. limit 2 projects 3 instead of 2/2

jburel commented:
FYI Point 1: this is difficult in the sense that for example all projects are loaded via api and not possible to verify in client e.g. log as "root"
Better to be tested one via "standard call" and one via api and compare the result

will-moore commented:
@jburel You can reproduce client behaviour of filtering by group & owner by using ```?group=3&owner=4```

jburel commented:
I was going for the general call. I think it will be better to rely on a test b/c I have to use filtering in order to validate. This means that I cannot validate w/o filtering

will-moore commented:
OK, I will add a test that checks the "totalCount" when filtering by group & owner.

jburel commented:
configuration parameters (``set omero.web.api.limit...``) are correctly taken into account
``
{"meta": {"totalCount": 1, "maxLimit": 5, "limit": 2, "offset": 0}
``

jburel commented:
If I specified a negative value for ``limit`` and ``offset`` it is set in ``meta``, this is not the case for ``MaxLimit`` and display me all the results
If negative values are passed maybe better to reset to the default

will-moore committed:
Fix 'settings' -> 'api_settings' in tests

will-moore committed:
Blitz E.g. -> e.g. fixes

will-moore committed:
More fixes for PR #5067

will-moore committed:
Test 'meta' 'totalCount' respects filtering by 'owner'

will-moore committed:
Test 'meta' 'totalCount' respects filtering by 'group'

will-moore committed:
Handle negative ?limit=-1 and ?offset=-1 use defaults instead

will-moore committed:
Use 'select count(distinct obj) to avoid duplicates

for example, duplicate Wells were being counted, 1 per WellSample.

will-moore commented:
Last commit fixes this issue:
When counting ```wells``` the query was counting Wells multiple times, once for each WellSample.

Wells are loaded with this query:
```
  select obj from Well obj join fetch obj.details.owner as owner join fetch obj.details.creationEvent
  left outer join fetch obj.wellSamples as wellSamples
  left outer join fetch wellSamples.image as image
  left outer join fetch wellSamples.plateAcquisition as plateAcquisition
  where obj.plate.id = :pid order by obj.column, obj.row, obj.id
```
Wells were being counted with this query:
```
select count(obj) from Well obj join  obj.details.owner as owner join  obj.details.creationEvent
left outer join  obj.wellSamples as wellSamples
left outer join  wellSamples.image as image
left outer join  wellSamples.plateAcquisition as plateAcquisition
where obj.plate.id = :pid 
```

atarkowska commented:
why do you need all optional joins if you don't use them in counting query?

will-moore commented:
For every query that we make with the /api/ we also do a 'count' query, so there is a simple conversion function that creates the 'count' query from the original 'data loading' query.
See ```conn.buildCountQuery()```.  This essentially does:
```
        query = query.replace("select obj ", "select count(distinct obj) ")
        query = query.replace("fetch", "")
        query = query.split("order by")[0]
```
Any more-complex conversion we do will have to work with ALL the queries, so will need to be more customised to different data types.

will-moore committed:
Check ['meta']['totalCount'] for api_project and api_wells tests

will-moore commented:
To test last few commits:
 - try negative numbers for ?offset and ?limit
 - Check ['meta']['totalCount'] json is correct for Wells in ```/plates/:id/wells/```

jburel commented:
``maxLimit`` and ``limit`` should be ``> 0`` e.g. ``?limit=0`` should reset to the default
I should not be able to set negative value in config for them
``localhost:4080/api/v0/m/projects/
``
result
``
{"meta": {"totalCount": 1, "maxLimit": -1, "limit": -1, "offset": 0}, "data": [{"url:project": "http://localhost:4080/api/v0/m/projects/1/", "omero:details": {"owner": {"UserName": "root", "FirstName": "root", "omero:details": {"@type": "TBD#Details", "permissions": {"canDelete": true, "perm": "------", "canEdit": true, "canAnnotate": true, "canLink": true, "@type": "TBD#Permissions"}}, "LastName": "root", "@id": 0, "@type": "http://www.openmicroscopy.org/Schemas/OME/2016-06#Experimenter"}, "group": {"omero:details": {"@type": "TBD#Details", "permissions": {"canDelete": true, "perm": "rw----", "canEdit": true, "canAnnotate": true, "canLink": true, "@type": "TBD#Permissions"}}, "@id": 0, "@type": "http://www.openmicroscopy.org/Schemas/OME/2016-06#ExperimenterGroup", "Name": "system"}, "@type": "TBD#Details", "permissions": {"canDelete": true, "perm": "rw----", "canEdit": true, "canAnnotate": true, "canLink": true, "@type": "TBD#Permissions"}}, "url:datasets": "http://localhost:4080/api/v0/m/projects/1/datasets/", "@id": 1, "@type": "http://www.openmicroscopy.org/Schemas/OME/2016-06#Project", "Name": "project1"}]}
``

will-moore commented:
If I go to https://cowfish.openmicroscopy.org/webtrial/api/v0/m/projects/?limit=-1
I see 
```
"meta": {
"totalCount": 64,
"maxLimit": 500,
"limit": 200,
"offset": 0
},
```

jburel commented:
That's not the problem.
I was allowed to set the configuration to negative value
e.g.
``
bin/omero config set omero.web.api.max_limit -1
``

will-moore commented:
If I use ?limit=0 (which is supported by the OMERO api) I get this, which is potentially useful
```
"meta": {
"totalCount": 64,
"maxLimit": 500,
"limit": 0,
"offset": 0
},
"data": []
```

will-moore commented:
Ah, OK - sorry. Yes, I guess the smallest ```max_limit``` should be 1

jburel commented:
It is useful if do not have any results
but when I do ``http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/1902/wells/?limit=0``
``
{"meta": {"totalCount": 10, "maxLimit": 500, "limit": 0, "offset": 0}, "data": [{"WellSamples": [{"omero:details": {"owner": {"UserName": "root", "FirstName": "root", "MiddleName": "", "omero:details": {"@type": "TBD#Details", "permissions": {"isUserWrite": false, "isWorldWrite": false, "canDelete": true, "isWorldRead": false, "perm": "------", "canEdit": true, "canAnnotate": true, "isGroupAnnotate": false, "isGroupWrite": false, "canLink": true, "isUserRead": false, "@type": "TBD#Permissions", "isGroupRead": false}}, "Email": "", "LastName": "root", "@id": 0, "@type": "http://www.openmicroscopy.org/Schemas/OME/2016-06#Experimenter"}, "group": {"omero:details": {"@type": "TBD#Details", "permissions": {"isUserWrite": true, "isWorldWrite": false, "canDelete": true, "isWorldRead": false, "perm": "
``
I still have data

will-moore commented:
Very strange - I can repeat locally too. But I think this must be something about that particular query, (loading Wells and joining WellSamples etc) that is causing limit=0 to be interpreted as limit=1.
It seems to work fine for other queries, e.g.
http://web-dev-merge.openmicroscopy.org/api/v0/m/projects/1226/datasets/?limit=0

jburel commented:
it works for projects, datasets etc.

will-moore committed:
Move 'offset' and 'limit' validation to validate_opts()

will-moore committed:
Remove try/except from gateway.util.getIntOrDefault()

Having this fail silently and return 0 is not helpful. Users need to know
that something has gone wrong.
Now uses same logic as webcient views.get_long_or_default()

will-moore committed:
Add test_invalid_parameter() to test_api_errors.py

will-moore committed:
tests include limit=-1 and limit=0

will-moore commented:
@joshmoore Any idea why using params of ```limit=0``` and ```offset=0``` for this query doesn't return 0 objects (I get 1 result):
```
select obj from Well obj join fetch obj.details.owner as owner join fetch obj.details.creationEvent
left outer join fetch obj.wellSamples as wellSamples
left outer join fetch wellSamples.image as image
left outer join fetch wellSamples.plateAcquisition as plateAcquisition
where obj.plate.id = :pid order by obj.column, obj.row, obj.id
```
When I don't load WellSamples etc then ```limit=0``` works OK (get no results) with
```
select obj from Well obj join fetch obj.details.owner as owner join fetch obj.details.creationEvent
where obj.plate.id = :pid order by obj.column, obj.row, obj.id
```

@jburel I now handle config of ```omero.web.api.limit``` and ```omero.web.api.max_limit``` to be a minimum value of 1.

joshmoore commented:
> @joshmoore Any idea why using params of limit=0 and offset=0 for this query doesn't return 0 objects (I get 1 result):

This might be caused by https://hibernate.atlassian.net/browse/HHH-8989 -- can you detect when `limit==0` and not run the query?

jburel commented:
The negative values are now correctly handled
@will-moore did you have a chance to look at @joshmoore's suggestion?

will-moore commented:
Sorry - missed @joshmoore's comment notification somehow. Yes, that should be trivial to do.

will-moore committed:
If ?limit=0 result = []

jburel commented:
Looks good Thanks all
